internal/pkg/cli/deploy/workload.go

```diff
@@ -365,9 +364,14 @@ func (d *workloadDeployer) uploadContainerImages(out *UploadArtifactsOutput) err
 	if len(buildArgsPerContainer) == 0 {
 		return nil
 	}
+	uri, err := d.repository.Login()
+	if err != nil {
+		return fmt.Errorf("login to docker: %w", err)
+	}
 	out.ImageDigests = make(map[string]ContainerImageIdentifier, len(buildArgsPerContainer))
 	for name, buildArgs := range buildArgsPerContainer {
-		digest, err := d.imageBuilderPusher.BuildAndPush(dockerengine.New(exec.NewCmd()), buildArgs)
+		buildArgs.URI = uri
```

Comment:
I might be missing something, but why do we need to pass `URI` as a build args 🤔 ? It looks like `URI` is stored as an internal state of `repository`

Reply:
So when `Login` is called we already making an [api call](https://github.com/aws/copilot-cli/pull/4697/files#diff-c3341fa47b7006ee5fa20ac85575594903aa23154c1427d2daa666e074d87e75R93) to fetch the `URI` of the repository. So if we set `URI` in `buildArgs` then we do not have to again make an [api call](https://github.com/aws/copilot-cli/blob/mainline/internal/pkg/repository/repository.go#L54) to fetch the URI.

Reply:
But we won't make the additional API call because `uri` is stored in `repository`, right? https://github.com/aws/copilot-cli/blob/67db7a2d4f94c69c836586f714341e587b1b8220/internal/pkg/repository/repository.go#L84

Reply:
```
// URI returns the uri of the repository.
func (r *Repository) URI() (string, error) {
	if r.uri != "" {
		return r.uri, nil
	}
	uri, err := r.registry.RepositoryURI(r.name)
	if err != nil {
		return "", fmt.Errorf("get repository URI: %w", err)
	}
	return uri, nil
}
```
Maybe we should set `r.uri = uri` before returning 🤔

Reply:
Done. changed the logic to set uri in `repository`.